dds/DCPS/SporadicTask.h

```diff
@@ -78,113 +83,72 @@ class SporadicTask : public RcEventHandler {
   virtual void execute(const MonotonicTimePoint& now) = 0;
 
 private:
-  WeakRcHandle<ReactorInterceptor> interceptor_;
-  bool scheduled_;
-  enum {
-    SPORADIC_TASK_NONE,
-    SPORADIC_TASK_SCHEDULE,
-    SPORADIC_TASK_CANCEL
-  } last_command_;
-  MonotonicTimePoint next_time_;
-  mutable ACE_Thread_Mutex mutex_;
-
-  struct ScheduleCommand : public ReactorInterceptor::Command {
-    ScheduleCommand(SporadicTask* hb, const TimeDuration& delay)
-      : sporadic_task_(hb), delay_(delay)
-    { }
-
-    virtual bool should_execute() const
-    {
-      return sporadic_task_->last_command_ != SPORADIC_TASK_SCHEDULE;
-    }
-
-    virtual void will_execute()
-    {
-      sporadic_task_->last_command_ = SPORADIC_TASK_SCHEDULE;
-    }
-
-    virtual void execute()
-    {
-      sporadic_task_->schedule_i(delay_);
-    }
-
-    virtual void queue_flushed()
-    {
-      sporadic_task_->last_command_ = SPORADIC_TASK_NONE;
-    }
-
-    SporadicTask* const sporadic_task_;
-    const TimeDuration delay_;
-  };
-
-  struct CancelCommand : public ReactorInterceptor::Command {
-    CancelCommand(SporadicTask* hb)
+  struct SporadicCommand : public ReactorInterceptor::Command {
+    SporadicCommand(SporadicTask* hb)
       : sporadic_task_(hb)
     { }
 
-    virtual bool should_execute() const
-    {
-      return sporadic_task_->last_command_ != SPORADIC_TASK_CANCEL;
-    }
-
-    virtual void will_execute()
-    {
-      sporadic_task_->last_command_ = SPORADIC_TASK_CANCEL;
-    }
-
     virtual void execute()
     {
-      sporadic_task_->cancel_i();
-    }
-
-    virtual void queue_flushed()
-    {
-      sporadic_task_->last_command_ = SPORADIC_TASK_NONE;
+      sporadic_task_->execute_i();
     }
 
     SporadicTask* const sporadic_task_;
   };
 
+  WeakRcHandle<ReactorInterceptor> interceptor_;
+  bool desired_scheduled_;
+  MonotonicTimePoint desired_next_time_;
+  TimeDuration desired_delay_;
+  bool actual_scheduled_;
+  MonotonicTimePoint actual_next_time_;
+  RcHandle<SporadicCommand> sporadic_command_;
+  mutable ACE_Thread_Mutex mutex_;
+
   int handle_timeout(const ACE_Time_Value& tv, const void*)
   {
     const MonotonicTimePoint now(tv);
     {
       ACE_Guard<ACE_Thread_Mutex> guard(mutex_);
-      scheduled_ = false;
+      desired_scheduled_ = false;
+      actual_scheduled_ = false;
     }
     execute(now);
     return 0;
   }
 
-  void schedule_i(const TimeDuration& delay)
+  void execute_i()
   {
+    // Bump up our refcount so the Reactor doesn't delete us.
+    RcHandle<SporadicTask> x = rchandle_from(this);
```

Comment:
Isn't deriving from `RcEventHandler` (which this class does) supposed to avoid this problem?  In order to run code in this method something else should already have a reference.  

Is this only for the case where a `cancel_timer` removes it from the reactor and that was the last reference?  If so, it would be nice for the reactor to take care of this, but maybe not...

Reply:
So there's a stack like this: `run_reactor_event_loop` -> `our_handler::handle_timeout` -> `cancel_timer`.  Wouldn't (or shouldn't) the `run_reactor_event_loop` level have its own reference?

Reply:
Yes.  The problematic stack is `run_reactor_event_loop` -> `not_our_handler::handle_timeout` -> `cancel_timer`.  Since it is running in our reactor, it executes immediately which may drop the reference count.